### PR TITLE
[STACK-1151 and STACK-1278] VRID Floating IP 

### PR DESCRIPTION
**Pull request by OmkarTelee-A10**
## Description
VRID Floating IP configuration
As an engineer, I can define what VRID Floating IP a project should use.
Depending on member subnet, engineer can provide the what Floating IP should be configured when a specific member create request has been received.

It requires configuration at 2 level.
1. Neutron level - create a port for Linuxbridge traffic forward 
2. Thunder level - AXAPI call to configure VRID

**Requirement** - Engineer is expected to enable VRRP-A settings before configuring VRID Floating IP.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1278
https://a10networks.atlassian.net/browse/STACK-1151

## Technical Approach
Created a VRID model, and database Table. This table tracks one:one relation between project and VRID floating IP. Written a task that handles VRID delta, (if changed in between creation of SLB or members). The delta includes creation and deletion of neutron port in member subnet and making AXAPI calls to configure specified VRID on device. This task return PORT model, based on port, we enter the updates in DB.
This DB entry acts as state management of VRID, and for future requests serves required info.

The provided VRID is checked for valid values like dhcp, whether the provided IP is in member subnet or not, and partial IP octets. The required functions are written in utils.

**Limitation**: Currently only VRID 0 is configurable for devices, if required in future, other than 0 can be easily implemented with help of this code

## Config Changes
We enable  engineer to provide VRIP Floating IP configurations at 2 places. 
1. [a10_global] section
2. [devices] section

Engineer can specify VRID Floating IP in a10_global if ALL of the members of device are in SAME member subnet.
Otherwise engineer can specify project wise VRID in devices section.
Expected values -
a. Actual VRID IP
b. 'DHCP' - One of the available IP will be used from subnet
c. Partial IP octet string - For ex. '8', if your subnet is 192.168.0.X, IP 192.168.0.8 will be configured as VRID Floating IP address

**Example CONFIG**

[a10_global]
enable_hierarchical_multitenancy = False
use_parent_partition = False
**vrid_floating_ip = '10.0.12.45'**

[vthunder]
default_vthunder_username = "admin"
default_vthunder_password = "a10"
default_axapi_version = "30"

[a10_controller_worker]
amp_image_owner_id = 20ceb94f69ba4407b646819fa2874565
amp_secgroup_list = lb-mgmt-sec-grp
amp_flavor_id = f2743f0d-e3d0-4a70-ad6c-f5302a8b80ab
amp_boot_network_list = 174a732a-41f8-451a-8492-1462afa7b053
amp_ssh_key_name = octavia_ssh_key
network_driver = a10_octavia_neutron_driver
workers = 2
amp_active_retries = 100
amp_active_wait_sec = 2
amp_image_id = 8fbf5836-19b9-4708-b101-6c737709af6d
loadbalancer_topology = SINGLE

[a10_health_manager]
udp_server_ip_address = 172.17.20.138
bind_port = 5550
bind_ip = 192.168.0.7
heartbeat_interval = 5
heartbeat_key = insecure
heartbeat_timeout = 90
health_check_interval = 3
failover_timeout = 600
health_check_timeout = 3
health_check_max_retries = 5

[a10_house_keeping]
load_balancer_expiry_age = 3600
amphorae_expiry_age = 3600

[LISTENER]
no-dest-nat = True

[hardware_thunder]
devices = [
 {
 "project_id": "eccaddd60f404477a7c32270267729d4",
 "username" : "admin",
 "password" : "a10",
 "ip_address" : "10.0.0.78",
 "device_name" : "rack_1",
 **"vrid_floating_ip" : "10.0.12.12"**
 },
 {
 "project_id": "076005c89eaf4983a36e548c4fa26152",
 "username" : "admin",
 "password" : "a10",
 "ip_address" : "10.0.0.78",
 "device_name" : "rack_1",
 "partition_name": "PartitionA",
 **"vrid_floating_ip" : "dhcp"**
 },
 {
 "project_id": "9cd6f31ab17449db982a26e70326aa50",
 "username" : "admin",
 "password" : "a10",
 "ip_address" : "10.0.0.78",
 "device_name" : "rack_1",
 "partition_name" : "PartitionB"
 }]

**NOTE:** If both global and project specific VRID Floating IP provided, project specific VRID is prefered

## Test Cases
All the test cases are added in below sheet 
https://a10networks.sharepoint.com/:x:/s/OpenstackTeam/EYYWK-LSxGpPpZoqKkLtVmUBSpu9KVlu5Ix96OpBtL2SNg?e=MtC7Ov

## Manual Testing
Follow this document to setup dev environment:
https://a10networks.sharepoint.com/:w:/s/OpenstackTeam/EYQMar32UuZKsWhVJ0ZJxG8Bgg_-8n570hXWne2HYbBhAw?e=W2tsbz

### CODE REVIEW

File: a10_octavia/tests/unit/controller/worker/tasks/test_a10_network_tasks.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+            subnet.network_id, member.subnet_id, fixed_ip=a10constants.MOCK_VRID_FLOATING_IP_1)
+        self.client_mock.vrrpa.update.assert_called_with(
+            0, floating_ip=a10constants.MOCK_VRID_FLOATING_IP_1)
+
+    # 2.b (SPECIFIC PARTITION) NO VRID, FIP provided
+    @mock.patch('a10_octavia.common.utils.get_vrid_floating_ip_for_project',
+                return_value=a10constants.MOCK_VRID_FLOATING_IP_1)
+    @mock.patch('a10_octavia.common.utils.get_patched_ip_address',
+                return_value=a10constants.MOCK_VRID_FLOATING_IP_1)
+    def test_handle_vrid_fip_with_no_vrid_but_fip_given_in_specific_partition(
```
Comment (hthompson-a10) [inline a10_octavia/tests/unit/controller/worker/tasks/test_a10_network_tasks.py:89]: If we have to decide between test names being about inputs or expected outcomes, we should tend towards the expected outcomes

Ideally `test_ClassName/function_name_operation_input` or `test_ClassName/function_name_input_operation`

`test_HandleVRIDFloatingIP_create_floating_ip_in_specified_partition_with_static_ip`

`test_HandleVRIDFloatingIP_with_static_ip_and_specified_parition_create_floating_ip_in_specified_partition`

There seems to be some divide on test names. But the main consensus is that they should tell you exactly what your testing, what goes in, what _should_ come out.

I think for most of our test cases `test_ClassName/function_name_operation_input` will look a little cleaner and be a bit shorter

File: a10_octavia/tests/unit/controller/worker/tasks/test_a10_network_tasks.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        mock_network_task.execute(vthunder, member, None)
+        self.network_driver_mock.create_port.assert_called_with(
+            subnet.network_id, member.subnet_id, fixed_ip=a10constants.MOCK_VRID_FLOATING_IP_1)
+        self.client_mock.vrrpa.update.assert_called_with(
+            0, floating_ip=a10constants.MOCK_VRID_FLOATING_IP_1, is_partition=True)
+
+    # 3.a (SHARED_PARTITION) NO VRID, provided "DHCP" as FIP
+    @mock.patch('a10_octavia.common.utils.check_ip_in_subnet_range', return_value=False)
+    @mock.patch('a10_octavia.common.utils.get_vrid_floating_ip_for_project', return_value='dhcp')
+    def test_handle_vrid_fip_with_no_vrid_but_dhcp_fip_given_in_shared_partition(
```
Comment (hthompson-a10) [inline a10_octavia/tests/unit/controller/worker/tasks/test_a10_network_tasks.py:112]: Another example
`test_HandleVRIDFloatingIP_create_floating_ip_in_shared_partition_with_dhcp`

File: a10_octavia/controller/worker/tasks/a10_network_tasks.py [hunk 3]
```diff
@@ -658,3 +661,86 @@ def revert(self, result, loadbalancer, amp_data=None, update_dict=None,
...
+        self.fip_port = None
+        super(HandleVRIDFloatingIP, self).__init__(*arg, **kwargs)
+
+    @axapi_client_decorator
+    def execute(self, vthunder, member, vrid):
+        vrid_ip = None
+        if vrid:
+            vrid_ip = vrid.vrid_floating_ip
+
+        floating_ip = a10_utils.get_vrid_floating_ip_for_project(member.project_id)
```
Comment (ssrinivasan-a10) [inline a10_octavia/controller/worker/tasks/a10_network_tasks.py:679]: May be we rename vrid_ip and floating_ip to db_floating_ip and conf_floating_ip.

File: a10_octavia/tests/unit/common/test_utils.py [hunk 4]
```diff
@@ -202,3 +216,71 @@ def test_merge_host_and_network_ip_valid(self):
...
         self.assertRaises(Exception, utils.merge_host_and_network_ip, '10.10.10.0/42', '99')
+
+    def test_get_patched_ip_address(self):
+        self.assertEqual(utils.get_patched_ip_address('45', '10.10.0.0/24'), '10.10.0.45')
+        self.assertEqual(utils.get_patched_ip_address('0.45', '10.10.0.0/24'), '10.10.0.45')
+        self.assertEqual(utils.get_patched_ip_address('0.0.45', '10.10.0.0/24'), '10.10.0.45')
+        self.assertEqual(utils.get_patched_ip_address('1.0.0.23', '10.10.0.0/24'), '1.0.0.23')
+
+    def test_get_patched_ip_address_invalid(self):
+        self.assertRaises(Exception, utils.get_patched_ip_address, '.45', '10.10.0.0/24')
```
[the remark this answers is not in the document]
Comment (afrin-chakure-a10) [reply a10_octavia/tests/unit/common/test_utils.py:227]: Fixed this bug

File: a10_octavia/tests/unit/common/test_utils.py [hunk 5]
```diff
@@ -202,3 +216,71 @@ def test_merge_host_and_network_ip_valid(self):
...
+        self.assertEqual(utils.get_vrid_floating_ip_for_project('project-2'), None)
+
+    def test_get_vrid_floating_ip_for_project_with_only_global_config(self):
+        vthunder = copy.deepcopy(VTHUNDER_1)
+        vthunder.project_id = a10constants.MOCK_PROJECT_ID
+        hardware_device_conf = [{"project_id": a10constants.MOCK_PROJECT_ID,
+                                 "username": vthunder.username,
+                                 "password": vthunder.password,
+                                 "device_name": vthunder.device_name,
+                                 "vrid_floating_ip": vthunder.vrid_floating_ip,
```
[the remark this answers is not in the document]
Comment (afrin-chakure-a10) [reply a10_octavia/tests/unit/common/test_utils.py:258]: Hardcoded it as None

**Pull request merged**